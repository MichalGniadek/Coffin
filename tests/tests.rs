use coffin2::{debug_print, lexer, name_resolution, parser, type_resolution};
use insta::{assert_snapshot, glob};
use std::{fs, path::Path};

fn test_util(path: &Path, spans: bool, variables: bool, types: bool) -> String {
    let code = fs::read_to_string(path).unwrap();
    let lexer = lexer::lex(&code);
    let mut ast = parser::parse(lexer);

    let variables = if variables | types {
        Some(name_resolution::visit(&mut ast))
    } else {
        None
    };

    let types = if types {
        Some(type_resolution::visit(
            &mut ast,
            variables.as_ref().unwrap(),
        ))
    } else {
        None
    };

    // let _module = spirv_generation::visit(&mut ast, &vars, &type_table);

    debug_print::visit(&ast, true, spans, variables.as_ref(), types.as_ref())
}

#[test]
fn parser() {
    glob!(r"insta_parser\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, false));
    });
}

#[test]
fn parser_spans() {
    glob!(r"insta_parser_spans\*.coff", |path| {
        assert_snapshot!(test_util(path, true, false, false));
    });
}

#[test]
fn variables() {
    glob!(r"insta_variables\*.coff", |path| {
        assert_snapshot!(test_util(path, false, true, false));
    });
}

#[test]
fn types() {
    glob!(r"insta_types\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, true));
    });
}
