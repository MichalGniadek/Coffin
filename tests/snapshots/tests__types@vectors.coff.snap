---
source: tests/tests.rs
expression: "test_util(path, false, false, true)"

---
unif img@0 : image2d@T17
fun f@1 (a@2 : float4@T16)  {
	($a@2: vector).xyzw
	($a@2: vector).xxx
	($a@2: vector).yx
	($a@2: vector).w
	($a@2: vector)[#3: int]
	($a@2: vector)[#0: int]
	(($img@0: image)[#123: int] = $a@2: vector): void
	($img@0: image)[#123: int]
	($a@2: vector).xyaw
	($a@2: vector).b
	($a@2: vector)[#4: int]
	($a@2: vector)[$a@2: vector]
	($a@2: vector).x.x
	(($a@2: vector).x = ($a@2: vector).y): void
	($img@0: image).asd
}: error

===
Mismatched type. Expected: vector, got: int. [114..117]
Mismatched type. Expected: vector, got: int. [134..137]
Incorrect vector fields. [148..152]
Incorrect vector fields. [160..161]
Mismatched type. Expected: int, got: vector. [179..180]
Swizzle not at the end. [191..192]
Type doesn't have fields. [223..226]
